core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,714 @@
+				     unsigned int idx, bool secure __unused)
+{
+	struct mmu_pgt *pgt = NULL;
+	struct mmu_pte *pte = NULL;
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	unsigned long ptp = 0;
+
+	if (tbl_info->level >= RISCV_PGLEVELS)
+		return false;
+
+	pgt = tbl_info->table;
+	pte = core_mmu_table_get_entry(pgt, idx);
+
+	if (core_mmu_entry_is_invalid(pte)) {
+		pgt = core_mmu_pgt_alloc(prtn);
+		if (!pgt)
+			return false;
+
+		ptp = core_mmu_ptp_create(pa_to_ppn((paddr_t)pgt));
+		core_mmu_entry_set(pte, ptp);
+	}
+
+	return true;
+}
+
+void core_mmu_set_info_table(struct core_mmu_table_info *tbl_info,
+			     unsigned int level, vaddr_t va_base, void *table)
+{
+	tbl_info->level = level;
+	tbl_info->table = table;
+	tbl_info->va_base = va_base;
+	tbl_info->shift = CORE_MMU_SHIFT_OF_LEVEL(level);
+	assert(level < RISCV_PGLEVELS);
+	tbl_info->num_entries = RISCV_PTES_PER_PT;
+}
+
+void core_mmu_get_entry_primitive(const void *table, size_t level,
+				  size_t idx, paddr_t *pa, uint32_t *attr)
+{
+	struct mmu_pgt *pgt = (struct mmu_pgt *)table;
+	struct mmu_pte *pte = core_mmu_table_get_entry(pgt, idx);
+
+	if (core_mmu_entry_is_valid(pte)) {
+		if (pa)
+			*pa = pte_to_pa(pte);
+		if (attr)
+			*attr = pte_to_mattr(level, pte);
+	} else {
+		if (pa)
+			*pa = 0;
+		if (attr)
+			*attr = 0;
+	}
+}
+
+void core_mmu_set_entry_primitive(void *table, size_t level, size_t idx,
+				  paddr_t pa, uint32_t attr)
+{
+	struct mmu_pgt *pgt = (struct mmu_pgt *)table;
+	struct mmu_pte *pte = core_mmu_table_get_entry(pgt, idx);
+	uint8_t perms = mattr_to_perms(level, attr);
+
+	core_mmu_entry_set(pte, core_mmu_pte_create(pa_to_ppn(pa), perms));
+}
+
+static void set_user_va_idx(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt = NULL;
+	struct mmu_pte *pte = NULL;
+	unsigned int idx = 0;
+
+	pgt = core_mmu_get_root_pgt_va(prtn);
+
+	for (idx = 1 ; idx < RISCV_PTES_PER_PT; idx++) {
+		pte = core_mmu_table_get_entry(pgt, idx);
+		if (core_mmu_entry_is_invalid(pte)) {
+			user_va_idx = idx;
+			return;
+		}
+	}
+	if (user_va_idx < 0)
+		panic();
+}
+
+static struct mmu_pte *
+core_mmu_get_user_mapping_entry(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt = core_mmu_get_root_pgt_va(prtn);
+
+	assert(user_va_idx != -1);
+
+	return core_mmu_table_get_entry(pgt, user_va_idx);
+}
+
+void core_mmu_set_user_map(struct core_mmu_user_map *map)
+{
+	unsigned long satp = 0;
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	struct mmu_pte *pte = NULL;
+	unsigned long ptp = 0;
+
+	satp = read_satp();
+	pte = core_mmu_get_user_mapping_entry(prtn);
+	if (map && map->user_map) {
+		ptp = core_mmu_ptp_create(pa_to_ppn((paddr_t)map->user_map));
+		core_mmu_entry_set(pte, ptp);
+		core_mmu_table_write_barrier();
+		satp |= SHIFT_U64(map->asid, RISCV_SATP_ASID_SHIFT);
+		tlbi_all();
+		write_satp(satp);
+	} else {
+		core_mmu_entry_set(pte, 0);
```

Comment:
Need `tlbi_all()` too after clearing PTE ?

`tlbi_all();` can be put after this if-else.